sycl/doc/extensions/experimental/sycl_ext_oneapi_device_global.asciidoc

```diff
@@ -615,18 +615,18 @@ global variable is allocated for each device, and that instance is shared by
 all kernels that belong to the same context and are submitted to the same device,
 regardless of which _device image_ contains the kernel.  When this property is
 specified, it is an assertion by the user that the device global is referenced
-only from kernels that are contained by the same _device image_.  An
+only from kernels that are contained by the same _device image_ instance, a _device image_ loaded onto a specific device. An
```

Comment:
I think you made this change to address this point in the PR description:

> 1. The spec doesn't currently state that all accesses to the device_global from device code must be from the same device.
>    * In the JiT flow you can build a device_image containing a device_global and then run kernels from that device_image on multiple targets. By the current wording of the spec, we were worried that, even with device_image_scope, a user could expect that device_global to retain its value across device.

I think this is not a valid concern, and I'd rather not make the change you suggest here.

The first paragraph of the section "[Representation of device globals](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_device_global.asciidoc#representation-of-device-globals)" says:

> `device_global` provides a mechanism to allocate device scope memory - memory which has unique underlying storage (of type T) for each `sycl::device` and `sycl::context` combination. If multiple valid device and context combinations are present then each receives its own unique underlying allocation. All kernels that reference the same `device_global` entity (either directly or via a pointer to its underlying object of type T) share the same allocation of that object when those kernels run on the same device and context.

All `device_global` variables store values that are scoped _to a particular device_, and this paragraph is attempting to state that.  This is true regardless of whether the device global uses the `device_image_scope` property.

While reading this, though, I think there is a different problem with the spec.  The paragraph I quoted above says that the storage of a device global variable is unique for each combination of `device` and `context`.  I think the part about `context` is not true when using the `device_image_scope` property.  In that case, I think the storage is the same _regardless_ of `context`.

Even when not using `device_image_scope`, I think that paragraph is not quite right regarding `context`.  Instead, I think the implementation has a requirement that all queues that access a device global variable must use the same `context`.  Tagging @steffenlarsen for his opinion here.

It occurs to me now that this is an unfortunate limitation, which could make it difficult to use device global variables in some programs.  I think the limitation comes from the requirement that USM device memory may only be accessed  from queues that have the same context as the context used to allocate that USM memory.  I wonder if there should be some way to allocate USM device memory that can be accessed from any context that is valid for the device on which it is allocated?  Tagging @bashbaug for his thoughts here.

Reply:
> which could make it difficult to use device global variables in some programs.

I think the proper "fix" for this issue is to minimize the number of contexts per device in a program, ideally down to just one, unless an application chooses to explicitly opt-in to multiple contexts.  We're already doing this with things like the "default context" that make it harder to accidentally create more contexts than intended.

Reply:
Sure, this is a good BKM.  However, the problem remains for applications that opt-in to using multiple contexts.  For these applications, it will be hard to use device global variables.

Is there a good reason to have such strict context rules for USM *device* memory?  If this type of memory is allocated on device *A* with context *C1*, is it really necessary that a kernel running on device *A* must also be launched with context *C1* in order to access it?  Regardless of context, it seems like a kernel running on device *A* should have access to any device memory allocated on that same device.  Is there really something in the hardware or low-level software stacks that requires the contexts to be the same?

Reply:
> Is there really something in the hardware or low-level software stacks that requires the contexts to be the same?

I think this discussion is expanding beyond the scope of this PR, but in general: yes, the contexts need to be the same.

Reply:
Thanks, @bashbaug.  I agree it's a bit of a tangent, but this discussion points to an error in that first paragraph in "[Representation of device globals](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_device_global.asciidoc#representation-of-device-globals)" that I quote above.

I'll describe the correction I think we need to make, but we can do that in a different PR.  My overall point is that the change proposed in this PR is not necessary because of this paragraph, which already states that the value of a `device_global` is per-device.  I think the correct wording of that paragraph should be like this:

> `device_global` provides a mechanism to allocate device scope memory - memory which has unique underlying storage (of type _T_) for each `sycl::device`.  All kernels that reference the same `device_global` entity (either directly or via a pointer to its underlying object of type _T_) share the same allocation of that object when those kernels run on the same device.
>
> Accesses to `device_global` have an important limitation with respect to the `sycl::context`.  All accesses to a `device_global` variable from a specific device must use the same context.  To illustrate, consider a kernel  that is submitted to device _D_ using context _C_, where the kernel accesses a `device_global` variable _V_.  Any time that kernel is submitted to device _D_, it must be submitted using context _C_.  Any other kernels that access the same variable _V_ must also use context _C_ when submitted to device _D_.  Any copy operations to or from variable _V_ on device _D_ must also use context _C_.